modules/photo/src/ccm/utils.cpp

```diff
@@ -0,0 +1,100 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Author: Longbu Wang <wanglongbu@huawei.com.com>
+//         Jinheng Zhang <zhangjinheng1@huawei.com>
+//         Chenqi Shan <shanchenqi@huawei.com>
+
+#include "utils.hpp"
+
+namespace cv {
+namespace ccm {
+
+inline double gammaCorrection_(const double& element, const double& gamma)
+{
+    return (element >= 0 ? pow(element, gamma) : -pow((-element), gamma));
+}
+
+Mat gammaCorrection(const Mat& src, const double& gamma, Mat dst)
+{
+    return elementWise(src, [gamma](double element) -> double { return gammaCorrection_(element, gamma); }, dst);
+}
+
+Mat maskCopyTo(const Mat& src, const Mat& mask)
+{
+    Mat dst(countNonZero(mask), 1, src.type());
+    const int channel = src.channels();
+    auto it_mask = mask.begin<uchar>();
+    switch (channel)
+    {
+    case 1:
+    {
+        auto it_src = src.begin<double>(), end_src = src.end<double>();
+        auto it_dst = dst.begin<double>();
+        for (; it_src != end_src; ++it_src, ++it_mask)
+        {
+            if (*it_mask)
+            {
+                (*it_dst) = (*it_src);
+                ++it_dst;
+            }
+        }
+        break;
+    }
+    case 3:
+    {
+        auto it_src = src.begin<Vec3d>(), end_src = src.end<Vec3d>();
+        auto it_dst = dst.begin<Vec3d>();
+        for (; it_src != end_src; ++it_src, ++it_mask)
+        {
+            if (*it_mask)
+            {
+                (*it_dst) = (*it_src);
+                ++it_dst;
+            }
+        }
+        break;
+    }
+    default:
+        CV_Error(Error::StsBadArg, "Wrong channel!" );
+        break;
+    }
+    return dst;
+}
+
+Mat multiple(const Mat& xyz, const Mat& ccm)
+{
+    Mat tmp = xyz.reshape(1, xyz.rows * xyz.cols);
+    Mat res = tmp * ccm;
+    res = res.reshape(res.cols, xyz.rows);
+    return res;
+}
+
+Mat saturate(Mat& src, const double& low, const double& up)
+{
+    Mat dst = Mat::ones(src.size(), CV_8UC1);
+    MatIterator_<Vec3d> it_src = src.begin<Vec3d>(), end_src = src.end<Vec3d>();
+    MatIterator_<uchar> it_dst = dst.begin<uchar>();
+    for (; it_src != end_src; ++it_src, ++it_dst)
+    {
+        for (int i = 0; i < 3; ++i)
+        {
+            if ((*it_src)[i] > up || (*it_src)[i] < low)
+            {
+                *it_dst = 0;
+                break;
+            }
+        }
+    }
+    return dst;
+}
```

Comment:
inRange should be enough instead of it: https://docs.opencv.org/4.x/d2/de8/group__core__array.html#ga48af0ab51e36436c5d04340e036ce981